[inode.c] Reduce work under table lock

black-dragon74 opened the pull request:
If inode is the root inode we always maintain the ref count as 1. In
that case there is no need of calling the `__inode_ref`. We check if the
inode is root with ref outside of the lock and then call the function
conditionally.

Updates: https://github.com/gluster/glusterfs/issues/3173

Change-Id: I3150b32ce31ae9a5b1ab47805929ec0b0f03a477
Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 committed:
[inode.c] Reduce work under table lock (1/2)

If inode is the root inode we always maintain the ref count as 1. In
that case there is no need of calling the `__inode_ref`. We check if the
inode is root with ref outside of the lock and then call the function
conditionally.

Updates: #3173

Change-Id: I3150b32ce31ae9a5b1ab47805929ec0b0f03a477
Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 committed:
Fix check for root inode in `inode_new`

Change-Id: Ied9f3e81d13c4ce350db7d204c4abd25177329da
Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
If this seems right for `inode_ref`, I will do the same for `inode_unref`. Please review and guide me.

Note: Tests are failing due to unused variable, this PR is still a WIP.

black-dragon74 committed:
[inode.c] Reduce work under table lock (2/2)

Change-Id: I7ad2fc3997ca52ca687fbbaf137b11fe83a39153
Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
/recheck smoke

black-dragon74 commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
https://build.gluster.org/job/gh_centos7-regression/2105/

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -598,18 +594,24 @@ inode_t *
 inode_unref(inode_t *inode)
 {
     inode_table_t *table = NULL;
+    xlator_t *xl = THIS;
```
You don't need to fetch THIS -  only in line 607.

black-dragon74 replied:
I did it this way to entirely remove fetching `THIS` inside the critical section. `THIS` calls `__glusterfs_this_location` so it would be evaluated inside the lock itself if we call `__inode_unref(..., THIS)`. Is it okay to evaluate inside the lock?

mykaul replied:
Indeed, it'd be great to fetch it outside the lock, but you can do it just before entering the lock, not right away unconditionally - as you might be exiting the function (say, it's a root node) without needing it at all. Some of my ongoing work is to reduce the unconditional fetch of 'THIS' everywhere in our code.

mykaul replied:
This still need to be adjusted, if possible.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -619,15 +621,20 @@ inode_t *
 inode_ref(inode_t *inode)
 {
     inode_table_t *table = NULL;
+    xlator_t *xl = THIS;
```
You don't need to fetch THIS - only in line 634

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -699,6 +706,8 @@ inode_t *
+    gf_boolean_t root_with_ref = _gf_false;
 
     if (!table) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, 0,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, 0,
```

black-dragon74 replied:
Totally missed this family of optimizations. Fixed.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -802,6 +815,7 @@ inode_grep(inode_table_t *table, inode_t *parent, const char *name)
+    xlator_t *xl = THIS;
 
     if (!table || !parent || !name) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, EINVAL, LG_MSG_INVALID_ARG,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, EINVAL, LG_MSG_INVALID_ARG,
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -917,6 +938,15 @@ __is_root_gfid(uuid_t gfid)
+static gf_boolean_t
+__is_root_with_ref(inode_t *inode)
+{
+    if (__is_root_gfid(inode->gfid) && inode->ref)
```
```suggestion
    if (inode->ref && __is_root_gfid(inode->gfid))
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -941,6 +971,7 @@ inode_t *
+    xlator_t *xl = THIS;
 
     if (!table) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, 0,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, 0,
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -1084,6 +1127,7 @@ inode_link(inode_t *inode, inode_t *parent, const char *name, struct iatt *iatt)
+    xlator_t *xl = THIS;
 
     if (!inode) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -1133,6 +1183,7 @@ int
+    xlator_t *xl = THIS;
 
     if (!inode) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -1177,19 +1228,23 @@ int
+    gf_boolean_t is_root_inode = _gf_false;
 
     if (!inode) {
         gf_msg_callingfn(THIS->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```
```suggestion
        gf_msg_callingfn(xl->name, GF_LOG_WARNING, 0, LG_MSG_INODE_NOT_FOUND,
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -1393,12 +1451,17 @@ inode_parent(inode_t *inode, uuid_t pargfid, const char *name)
     }
     pthread_mutex_unlock(&table->lock);
 
+    if (parent && !__is_root_with_ref(parent)) {
```
Q: does it make sense to unlock the table and right away lock it if this condition is true? Unsure.

black-dragon74 replied:
I actually read about it before implementing it this way, it adds `80-90ns` overhead. IMO, the cases where this condition is false will compensate and overall balance this approach. What do you think?

mykaul replied:
Unsure really. Let's keep it the way you did it, we can always come back to it later. The latency we are concerned about is not how much the 'if' statement takes, nor how much we save by skipping the unlock, lock steps - it's how much contention the table lock has, and if by skipping this we actually complete some work faster - not always clear what's the winning strategy is.

xhernandez replied:
Besides the potential performance gains (which I think are very marginal if any at all), this creates a race which is more dangerous.

xhernandez replied:
This needs to be addressed.

mykaul commented:
My (incomplete) code is a bit more humble in the changes - see https://github.com/mykaul/glusterfs/commit/9a1fb1b56410dbb481e7632a55cc3ddeba71a997 - I think incremental changes would be easier to evaluate.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -464,18 +467,16 @@ __inode_get_xl_index(inode_t *inode, xlator_t *xlator)
     /*
      * Root inode should always be in active list of inode table. So unrefs
-     * on root inode are no-ops.
+     * on root inode are no-ops. Call __inode_unref(3) only when __is_root_gfid
```
```suggestion
     * on root inode are no-ops. Call __inode_unref() only when __is_root_gfid
```

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -523,9 +522,9 @@ __inode_unref(inode_t *inode, bool clear)
+    index = __inode_get_xl_index(inode, xl);
     if (index >= 0) {
-        inode->_ctx[index].xl_key = this;
+        inode->_ctx[index].xl_key = xl;
```
nit (and in a different patch) - I don't see why this line is needed at all.

black-dragon74 replied:
Noted. `xl_key` is already set to `xl` when `index >= 0`, so we can safely remove this line. Patch coming soon.

black-dragon74 replied:
Proposed at: #3207

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -619,15 +621,20 @@ inode_t *
     if (!inode)
         return NULL;
 
     table = inode->table;
```
this could be moved to line 634.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -955,17 +986,23 @@ inode_find(inode_table_t *table, uuid_t gfid)
     pthread_mutex_lock(&table->lock);
     {
         inode = __inode_find(table, gfid, hash);
```
Why do you need to hash (line 984) if it's the root?

black-dragon74 replied:
Nice catch.

xhernandez commented:
I don't think this is a good idea. It makes it easy to forget checking for root inode before calling `__inode_{ref,unref}()`, so we may actually release and destroy the root inode quite easily. That's too dangerous.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -815,14 +831,21 @@ inode_grep(inode_table_t *table, inode_t *parent, const char *name)
     }
     pthread_mutex_unlock(&table->lock);
 
+    if (dentry->inode) {
```
That's a bad idea. Once we leave the mutex, the `dentry` could have been destroyed by another thread, so we are accessing potentially invalid memory here. We cannot leave the mutex before taking a reference of the inode.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -941,31 +974,36 @@ inode_t *
     }
     pthread_mutex_unlock(&table->lock);
 
+    if (inode && !__is_root_with_ref((inode))) {
```
This also causes a potential race.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -1104,12 +1148,18 @@ inode_link(inode_t *inode, inode_t *parent, const char *name, struct iatt *iatt)
     }
     pthread_mutex_unlock(&table->lock);
 
+    if (linked_inode && !__is_root_with_ref(linked_inode)) {
```
Potential race.

black-dragon74 replied:
Taken care of. Thank you.

black-dragon74 commented:
I think opening a new PR with last 3 commits of this PR would be nice (as those are just simple optimizations). And then I can rebase it to defdcbe8082516af5c367a3ce0999d3523bb653f.

xhernandez commented:
> I don't think this is a good idea. It makes it easy to forget checking for root inode before calling `__inode_{ref,unref}()`, so we may actually release and destroy the root inode quite easily. That's too dangerous.

@black-dragon74 after a second look, I've seen that `__inode_{ref,unref}()` are `static`, so we only need to care about them inside `inode.c`, which limits the risks I commented. In this case I think it's ok to move the check outside of the mutex, but the remaining issues need to be fixed.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -710,13 +719,17 @@ inode_new(inode_table_t *table)
 
     inode = inode_create(table);
     if (inode) {
+        if (__is_root_with_ref(inode))
```
need to see how we can improve this - we create the root inode very very rarely. Checking it all the time for the root is somewhat inefficient. Perhaps in a different patch.

xhernandez replied:
> need to see how we can improve this - we create the root inode very very rarely. Checking it all the time for the root is somewhat inefficient. Perhaps in a different patch.

A new inode will always have 0 references. I guess we don't need to call this.

black-dragon74 replied:
Done.

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -917,6 +938,15 @@ __is_root_gfid(uuid_t gfid)
+static gf_boolean_t
+__is_root_with_ref(inode_t *inode)
+{
+    if (__is_root_gfid(inode->gfid) && inode->ref)
```
As opposed to the gfid, which is not going to change, can't the ref change, if you are testing it not under the lock?

xhernandez replied:
Except after creation, root inode should always have >= 1 references. After creation I think we can avoid calling this function, so the check is not needed.

black-dragon74 committed:
[inode.c] Optimize fetching THIS and fix races

Reuse `THIS` once fetched and fix some races that resulted due
to prior modifications. Also optimized the `__is_root_with_ref`
function to compare the scalar first.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index 316520545..56d6159d9 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -295,7 +295,8 @@ gf_gfid_generate_from_xxh64(uuid_t gfid, char *key)
     }
 
     gf_msg_debug(this->name, 0,
-                 "gfid generated is %s (hash1: %" XXH64_FMT ") "
+                 "gfid generated is %s (hash1: %" XXH64_FMT
+                 ") "
                  "hash2: %" XXH64_FMT ", xxh64_1: %s xxh64_2: %s",
                  uuid_utoa(gfid), hash_1, hash_2, xxh64_1, xxh64_2);
 
@@ -5466,7 +5467,7 @@ gf_pipe(int fd[2], int flags)
     int ret = 0;
 #if defined(HAVE_PIPE2)
     ret = pipe2(fd, flags);
-#else /* not HAVE_PIPE2 */
+#else  /* not HAVE_PIPE2 */
     ret = pipe(fd);
     if (ret < 0)
         return ret;
diff --git a/libglusterfs/src/gf-io-common.c b/libglusterfs/src/gf-io-common.c
index e10140995..55c109720 100644
--- a/libglusterfs/src/gf-io-common.c
+++ b/libglusterfs/src/gf-io-common.c
@@ -497,8 +497,8 @@ gf_io_thread_join(pthread_t thread, struct timespec *timeout)
 {
 #ifdef GF_LINUX_HOST_OS
     if (timeout != NULL) {
-        gf_io_success(gf_io_call_ret(pthread_timedjoin_np, thread, NULL,
-                                     timeout));
+        gf_io_success(
+            gf_io_call_ret(pthread_timedjoin_np, thread, NULL, timeout));
 
         return;
     }
diff --git a/libglusterfs/src/glusterfs/iobuf.h b/libglusterfs/src/glusterfs/iobuf.h
index 53b73d7a8..3e7172edf 100644
--- a/libglusterfs/src/glusterfs/iobuf.h
+++ b/libglusterfs/src/glusterfs/iobuf.h
@@ -69,8 +69,8 @@ struct iobuf {
 
     void *ptr; /* usable memory region by the consumer */
 
-    void *free_ptr;   /* in case of stdalloc, this is the
-                         one to be freed */
+    void *free_ptr; /* in case of stdalloc, this is the
+                       one to be freed */
     size_t page_size;
 };
 
diff --git a/xlators/mgmt/glusterd/src/glusterd-op-sm.c b/xlators/mgmt/glusterd/src/glusterd-op-sm.c
index b546dbf78..168c3acbe 100644
--- a/xlators/mgmt/glusterd/src/glusterd-op-sm.c
+++ b/xlators/mgmt/glusterd/src/glusterd-op-sm.c
@@ -131,8 +131,8 @@ glusterd_txn_opinfo_dict_fini()
 
 void
 glusterd_txn_opinfo_init(glusterd_op_info_t *opinfo,
-                         glusterd_op_sm_state_t state, int *op,
-                         dict_t *op_ctx, rpcsvc_request_t *req)
+                         glusterd_op_sm_state_t state, int *op, dict_t *op_ctx,
+                         rpcsvc_request_t *req)
 {
     glusterd_conf_t *conf = NULL;
 
@@ -5826,9 +5826,9 @@ glusterd_op_sm_transition_state(glusterd_op_info_t *opinfo,
     conf = THIS->private;
     GF_ASSERT(conf);
 
-    (void)glusterd_sm_tr_log_transition_add(
-        &conf->op_sm_log, opinfo->state, state[event_type].next_state,
-        event_type);
+    (void)glusterd_sm_tr_log_transition_add(&conf->op_sm_log, opinfo->state,
+                                            state[event_type].next_state,
+                                            event_type);
 
     opinfo->state = state[event_type].next_state;
     return 0;
diff --git a/xlators/mgmt/glusterd/src/snapshot/glusterd-lvm-snapshot.c b/xlators/mgmt/glusterd/src/snapshot/glusterd-lvm-snapshot.c
index 3e3c196f6..0e15d9efb 100644
--- a/xlators/mgmt/glusterd/src/snapshot/glusterd-lvm-snapshot.c
+++ b/xlators/mgmt/glusterd/src/snapshot/glusterd-lvm-snapshot.c
@@ -558,7 +558,7 @@ glusterd_lvm_brick_details(dict_t *rsp_dict,
         if (!token) {
             ret = -1;
             gf_msg(this->name, GF_LOG_ERROR, EINVAL, GD_MSG_INVALID_ENTRY,
-                    "Invalid vg entry");
+                   "Invalid vg entry");
             goto end;
         }
         value = gf_strdup(token);
@@ -574,7 +574,7 @@ glusterd_lvm_brick_details(dict_t *rsp_dict,
         ret = dict_set_dynstr(rsp_dict, key, value);
         if (ret) {
             gf_msg(this->name, GF_LOG_ERROR, 0, GD_MSG_DICT_SET_FAILED,
-                    "Could not save vgname ");
+                   "Could not save vgname ");
             goto end;
         }
 
diff --git a/xlators/mgmt/glusterd/src/snapshot/glusterd-zfs-snapshot.c b/xlators/mgmt/glusterd/src/snapshot/glusterd-zfs-snapshot.c
index 41f97caae..34d7ff944 100644
--- a/xlators/mgmt/glusterd/src/snapshot/glusterd-zfs-snapshot.c
+++ b/xlators/mgmt/glusterd/src/snapshot/glusterd-zfs-snapshot.c
@@ -488,9 +488,9 @@ glusterd_zfs_snap_clone_brick_path(char *snap_mount_dir,
                        snap_brick_dir);
     else {
         if (restore)
-            len = snprintf(brickinfo->path, sizeof(brickinfo->path), "%s/%s/brick%d%s",
-                           snap_mount_dir, snap_clone_volume_id, brick_num,
-                           snap_brick_dir);
+            len = snprintf(brickinfo->path, sizeof(brickinfo->path),
+                           "%s/%s/brick%d%s", snap_mount_dir,
+                           snap_clone_volume_id, brick_num, snap_brick_dir);
         else
             len = snprintf(brickinfo->path, sizeof(brickinfo->path),
                            "%s/.zfs/snapshot/%s_%d/%s", origin_brick_mount,
diff --git a/xlators/protocol/client/src/client-common.h b/xlators/protocol/client/src/client-common.h
index 996e2ed9d..b4bd7e363 100644
--- a/xlators/protocol/client/src/client-common.h
+++ b/xlators/protocol/client/src/client-common.h
@@ -18,7 +18,6 @@
 #include "glusterfs3.h"
 #include "client.h"
 
-
 /* New functions for version 4 */
 int
 client_post_common_dict(xlator_t *this, gfx_common_dict_rsp *rsp, dict_t **dict,
```

black-dragon74 commented:
Any idea why the warning of clang-format is coming up? I did not event touch those files in this patch 0_o

black-dragon74 committed:
Merge branch 'devel' into inode-lk-3173

black-dragon74 commented:
/recheck smoke-centos7

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -542,45 +542,41 @@ __inode_unref(inode_t *inode, bool clear)
-    index = __inode_get_xl_index(inode, this);
-    if (index >= 0) {
+    index = __inode_get_xl_index(inode, xl);
+    if (index >= 0)
```
Please, don't remove curly brackets. They are required based on our [coding standard](https://github.com/gluster/glusterfs/blob/devel/doc/developer-guide/coding-standard.md#always-use-braces).

black-dragon74 replied:
Yes, I remember reading this doc but then I have noticed at several places in our code where there is only a single statement inside the if block, curly braces are omitted, that led to the confusion. I guess they must be a part of old code. Anyways, got it clearly now. Thank you!

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -732,7 +741,8 @@ inode_new(inode_table_t *table)
             GF_ASSERT(!inode->in_lru_list);
             inode->in_lru_list = _gf_true;
-            __inode_ref(inode, false);
+            if (!root_with_ref)
```
Isn't this always true ?

black-dragon74 replied:
Missed it while removing the check for root's ref in the earlier commit. Fixed now.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -831,10 +842,13 @@ inode_grep(inode_table_t *table, inode_t *parent, const char *name)
         dentry = __dentry_grep(table, parent, name, hash);
-        if (dentry) {
+
+        if (dentry->inode) {
```
Can't `dentry` be `NULL` ?

black-dragon74 replied:
Fixed.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -933,6 +947,15 @@ __is_root_gfid(uuid_t gfid)
+static gf_boolean_t
+__is_root_with_ref(inode_t *inode)
+{
+    if (inode->ref && __is_root_gfid(inode->gfid))
```
Why we do need to check the number of references ? the root inode will always have 1 reference at least (except on creation, but this function is not used whem created).

black-dragon74 replied:
The previous code used to check for it and hence I also added the same. Following our discussions, I know that root inode has ref < 1 only inside `inode_new`. We are safe to assume that ref will be > 1 everywhere else. I am removing this function then as it serves no purpose.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -1651,17 +1701,18 @@ inode_table_prune(inode_table_t *table)
             if (!ret1) {
                 tmp->invalidate_sent = true;
-                __inode_unref(tmp, false);
+                (void)(!tmp_is_root && __inode_unref(tmp, false, old_THIS));
```
Bad way of writing code. Please use a regular `if` statement.

black-dragon74 replied:
Noted and addressed.

black-dragon74 committed:
[inode.c] Remove `is_root_with_ref` function

Root inode will always have a ref count of 1 except when
created. Hence, there is no need to check for it explicitly.

Signed-off-by: black-dragon74 <niryadav@redhat.com>

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -831,10 +839,22 @@ inode_grep(inode_table_t *table, inode_t *parent, const char *name)
+
+        if (!dentry) {
+            pthread_mutex_unlock(&table->lock);
+            gf_msg_callingfn(xl->name, GF_LOG_WARNING, EINVAL,
```
Not finding a dentry could be a "normal" issue in some cases, which means that we could flood the logs with this message. I wouldn't use `gf_msg_callingfn()` for this, which is costly. I would also use `GF_LOG_DEBUG` at most.

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -831,10 +839,22 @@ inode_grep(inode_table_t *table, inode_t *parent, const char *name)
-                __inode_ref(inode, false);
+        }
+
+        if (!__is_root_gfid(inode->gfid)) {
+            __inode_ref(inode, false, xl);
```
This must be done only inside the `if` of line 852. Otherwise `inode` can be `NULL`.

xhernandez reviewed (commented):
I see that in many cases `__is_root_gfid()` still needs to be called inside the mutex. So I'm wondering if it really makes sense to change this.

The benefit is limited to few cases where it can be done outside the mutex, but in all other cases it just makes the code less clean.

black-dragon74 committed:
[inode.c] Remove unnecessary logging and fixes

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
Smoke for centos is failing due to SSH timeout. I doubt that it is related to this patch.

black-dragon74 commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
https://build.gluster.org/job/gh_centos7-regression/2390/

black-dragon74 commented:
/recheck smoke
